common/djangoapps/util/db.py

```diff
@@ -4,44 +4,119 @@
+            raise transaction.TransactionManagementError('Cannot be inside an atomic block.')
+
+        if getattr(connection, 'commit_on_success_block_level', 0) == 0:
+            connection.commit_on_success_block_level = 1
+
+            # This will set the transaction isolation level to READ COMMITTED for the next transaction.
+            if self.read_committed is True:
+                if connection.vendor == 'mysql':
+                    cursor = connection.cursor()
+                    cursor.execute("SET TRANSACTION ISOLATION LEVEL READ COMMITTED")
+
+            # We aren't in a transaction yet; create one.
+            # The usual way to start a transaction is to turn autocommit off.
+            # However, some database adapters (namely sqlite3) don't handle
+            # transactions and savepoints properly when autocommit is off.
+            # In such cases, start an explicit transaction instead, which has
+            # the side-effect of disabling autocommit.
+            if connection.features.autocommits_when_autocommit_is_off:
+                connection._start_transaction_under_autocommit()
+                connection.autocommit = False
+            else:
+                connection.set_autocommit(False)
+        else:
+            connection.commit_on_success_block_level += 1
+
+    def __exit__(self, exc_type, exc_value, traceback):
+
+        if not self.ENABLED:
+            return
+
+        connection = transaction.get_connection(self.using)
+
+        try:
+            if exc_type is None:
+                # Commit transaction
+                try:
+                    connection.commit()
+                except DatabaseError:
+                    try:
+                        connection.rollback()
+                    except Error:
+                        # An error during rollback means that something
+                        # went wrong with the connection. Drop it.
+                        connection.close()
+                    raise
+            else:
+                # Roll back transaction
+                try:
+                    connection.rollback()
+                except Error:
+                    # An error during rollback means that something
+                    # went wrong with the connection. Drop it.
+                    connection.close()
+
+        finally:
+            connection.commit_on_success_block_level -= 1
+
+            # Outermost block exit when autocommit was enabled.
+            if connection.commit_on_success_block_level == 0:
+                if connection.features.autocommits_when_autocommit_is_off:
+                    connection.autocommit = True
+                else:
+                    connection.set_autocommit(True)
+
+    def __call__(self, func):
+        @wraps(func)
+        def decorated(*args, **kwds):
+            with self:
+                return func(*args, **kwds)
+        return decorated
+
+
+def commit_on_success(using=None, read_committed=False):
```

Comment:
I'm a +1 on this. I would like to see us discourage things like this in the platform and backlog a discovery item for us to come up with a plan to better deal with the situations that currently require us to do things such as this.